src/Documentation/Event-Sourcing/GrainStateAPI.md

```diff
@@ -0,0 +1,112 @@
+
+_Important:_ The application should never directly modify the object returned by `State`. It is meant for reading only. Rather, when the application wants to modify the state, it must do so indirectly by raising events.
+
+## Raising Events
+
+Raising events is accomplished by calling the `RaiseEvent` function. For example, a grain representing a chat can raise a `PostedEvent` to indicate that a user submitted a post:
+
+```csharp
+RaiseEvent(new PostedEvent() { Guid = guid, User = user, Text = text, Timestamp = DateTime.UtcNow });
+```
+
+Note that `RaiseEvent` kicks off a write to storage access, but does not wait for the write to complete. For many applications, it is important to wait until we have confirmation that the event has been persisted. In that case, we always follow up by waiting for `ConfirmEvents`:
+
+```csharp
+RaiseEvent(new DepositTransaction() { DepositAmount = amount, Description = description });
+await ConfirmEvents();
+```
+
+Note that even if you don't explicitly call `ConfirmEvents`, the events will eventually be confirmed - it happens automatically in the background. For more discussion on this topic, see [Immediate vs. Delayed Confirmation](MultiVersion.md).
+
+## State Transition Methods
+
+The runtime updates the grain state _automatically_ whenever events are raised. There is no need for the application to explicitly update the state after raising an event. However, the application still has to provide the code that specifies _how_ to update the state in response to an event. This can be done in two ways.
+
+**(a)** The GrainState class can implement one or more `Apply` methods on the `StateType`. Typically, one would create multiple overloads, and the closest match is chosen for the runtime type of the event:
+```csharp
+class GrainState {
+   
+   Apply(E1 @event)  
+   {
+     // code that updates the state
+   }
+   Apply(E2 @event)  
+   {
+     // code that updates the state
+   }
+}
+```
+
+**(b)** The grain can override the TransitionState function:
+```csharp
+protected override void TransitionState(State state, EventType @event)
+{
+   // code that updates the state
+}
+```
+
+The transition methods are assumed to have no side effects other than modifying the state object, and should be deterministic (otherwise, the effects are unpredictable).  If the transition code throws an exception, that exception is caught and included in a warning in the Orleans log, issued by the log-consistency provider.  
+
+When, exactly, the runtime calls the transition methods depends on the chosen log consistency provider and its configuration. It is best for applications not to rely on a particular timing, except when specifically guaranteed by the log consistency provider. 
```

Comment:
Does it mean that there could be race conditions if I raise an event, don't wait for it to be confirmed, and then start handling the next request?